File: guest/go/fkgc/meta.go
//go:build gc.custom

package fkgc

import "unsafe"

// ---------------------------------------------------------------------------
// Collector metadata that GROWS WITH THE HEAP, which is what removes the cap.
//
// Until sharding stage C the mark bitmap, the span table and the span-aux table
// were three fixed-size arrays inside the .bss `gcMeta` struct, and .bss is
// sized at link time -- so the heap they described had a compile-time maximum
// and `fkgc.HeapCap` was a HARD cap a guest trapped against. Three build tags
// moved it. All of that is deleted: FkLua imposes no memory cap beyond what
// Factorio imposes, and a collected guest now grows exactly like a leaking one.
//
// THE SHAPE. The metadata splits by GROWTH LAW rather than by what it means:
//
//	fixed      the class tables, the mark stack, the free-run heads, the dirty
//	           queue, the counters. ~28 KiB, still .bss, still exactly one
//	           struct (see gcMeta for why that is a correctness requirement).
//	scaling    the mark bitmap, spanClass and spanAux. One CHUNK per 4 MiB
//	           slice of heap, allocated FROM THE HEAP under the class clsMeta,
//	           reached through a directory that is itself a fixed .bss array.
//
// `metaDir` is 1,024 entries of 4 KiB of .bss and that covers the ENTIRE wasm32
// address space: 1,024 x 4 MiB is 4 GiB. A flat table sized for 4 GiB would
// have been 35 MiB of .bss -- about 7 ms of Factorio worst tick before the
// guest allocated a byte -- which is why the directory exists at all.
//
// WHY THIS IS SAFE, and it is the one hazard in the design with no precedent
// elsewhere in the repo. agents/gc.md section 6 put metadata outside the heap so
// that the collector writing its own bitmap would not dirty cards the next
// incremental step has to re-scan. Static placement is NOT what neutralises
// that, and sharding stage A measured why: `drainDirtySpans` already drops a
// sub-heap card with one compare, and a metadata card inside the heap is
// dropped by the span-class load `rescanSpan` ALREADY PERFORMS. The compare
// moves; it does not multiply. What static placement really bought was a bound,
// and the bound was the defect.
//
// Two things follow and both are enforced below rather than assumed:
//
//   - `markCandidate` REJECTS a clsMeta span, so no mark bit is ever set for a
//     metadata byte and the collector cannot retain its own bookkeeping as if
//     it were an object.
//   - `metaDir` holds HEAP ADDRESSES and lives inside [metaLo, metaHi), so
//     `scanRoots` subtracts it. A directory in an ordinary global would be
//     scanned as roots, and every chunk would be marked live through it.
//
// THE COST, stated rather than buried. Per MiB of heap the scaling part is
// 10,240 B (0.977%) against the old static scheme's 8,960 B (0.854%): the three
// tables are uint32 now rather than uint32/uint8/uint16, because a chunk is
// LINEAR MEMORY and a byte or halfword store there is a read-modify-write in
// emitted Lua while a word store is one table assignment. 14% more metadata to
// take a read-modify-write off the sweep's inner loop is the right side of that
// trade, and the exact model is asserted by TestTheMetadataSizeModelHolds
// rather than written down as prose that goes stale (it did: the shipped
// comments said 42 and 645 KiB against a measured 58.32 and 583.32).
// ---------------------------------------------------------------------------

const (
	// sliceLog is how much heap one chunk describes: 4 MiB.
	//
	// It is the whole tuning decision here and it trades two things against
	// each other. Bigger slices mean a smaller directory and a larger FLOOR --
	// the first chunk is paid by any guest that allocates at all. Smaller
	// slices mean the reverse. At 4 MiB the floor is 40 KiB and the directory
	// is 4 KiB; at 64 MiB the floor would be 573 KiB, which is worse than the
	// 163 KiB static scheme this replaces for every guest that never gets
	// there.
	sliceLog   = 22
	sliceSpans = 1 << (sliceLog - spanLog)    // 1,024 spans per chunk
	sliceWords = 1 << (sliceLog - granuleLog) // 262,144 granules per chunk

	// The chunk's layout, in bytes from its base. Three arrays of uint32, in
	// the order they are hottest.
	metaMarkOff  = 0
	metaMarkLen  = sliceWords / 32 * 4 // 32,768 B: one bit per granule
	metaClassOff = metaMarkOff + metaMarkLen
	metaClassLen = sliceSpans * 4 // 4,096 B: one word per span
	metaAuxOff   = metaClassOff + metaClassLen
	metaAuxLen   = sliceSpans * 4 // 4,096 B

	// 40,960 B is exactly ten spans, with nothing left over. That is not luck
	// -- it is what picking uint32 for all three tables buys, and a layout that
	// did not land on a span boundary would waste the remainder of the last
	// one forever.
	metaChunkBytes = metaAuxOff + metaAuxLen
	metaChunkSpans = metaChunkBytes >> spanLog

	// maxChunks covers all of wasm32. See maxSpans for the bound that actually
	// binds, which is smaller and is about uint32 arithmetic rather than about
	// this.
	maxChunks = 1 << (32 - sliceLog)

	// markWordsPerSpan is 8: a span is 4,096 B, a granule is 16 B, so a span is
	// exactly 256 granules and exactly eight bitmap words. The bitmap
	// partitions on span boundaries with nothing left over, which is what lets
	// a sweep clear as it goes and what lets every per-span loop below compute
	// ONE address and then index within it.
	markWordsPerSpan = spanBytes / granule / 32
)

// spanClassOf is what span si holds: 0 unassigned, 1..numClasses a size class,
// clsLarge/clsLargeMid a large-object run, clsMeta a metadata chunk.
//
// The value may carry clsFresh; every reader that cares about the CLASS masks
// it off, and sweepSpan is the one reader that wants the raw word. See clsFresh.
func spanClassOf(si uint32) uint32 {
	return load32(gcm.metaDir[si>>(sliceLog-spanLog)] + metaClassOff + (si&(sliceSpans-1))<<2)
}

func setSpanClass(si, c uint32) {
	store32(gcm.metaDir[si>>(sliceLog-spanLog)]+metaClassOff+(si&(sliceSpans-1))<<2, c)
}

// spanAuxOf is the run length for a clsLarge head and the head's span index for
// a clsLargeMid continuation. Unused otherwise.
func spanAuxOf(si uint32) uint32 {
	return load32(gcm.metaDir[si>>(sliceLog-spanLog)] + metaAuxOff + (si&(sliceSpans-1))<<2)
}

func setSpanAux(si, v uint32) {
	store32(gcm.metaDir[si>>(sliceLog-spanLog)]+metaAuxOff+(si&(sliceSpans-1))<<2, v)
}

// markWordBase is the address of the FIRST of span si's eight bitmap words.
//
// Every per-span loop in the sweep and the re-scan calls this once and then
// indexes within the span, which is why moving the bitmap into the heap did not
// make either of them slower: the old code recomputed (addr-heapBase)>>4 and
// indexed a global array per SLOT, and this computes one address per SPAN.
func markWordBase(si uint32) uint32 {
	return gcm.metaDir[si>>(sliceLog-spanLog)] + metaMarkOff +
		(si&(sliceSpans-1))*(markWordsPerSpan*4)
}

// isMarkedAt tests the bit for the object at byte offset off within the span
// whose bitmap words start at mw. off is a span offset, so it is under 4,096
// and the granule index is under 256.
func isMarkedAt(mw, off uint32) bool {
	g := off >> granuleLog
	return load32(mw+(g>>5)<<2)&(uint32(1)<<(g&31)) != 0
}

// clearSpanMarks wipes span si's eight bitmap words.
//
// Eight explicit stores rather than a clear() over a slice: clear() lowers to
// llvm.memset, which the emitter substitutes with the runtime's mem_fill -- a
// call with its own bounds check and page mark, for 32 bytes. Stage B measured
// that shape and the finding was the same one (see freeInvariant): the CALL is
// the cost, not the bytes.
func clearSpanMarks(si uint32) {
	w := markWordBase(si)
	store32(w, 0)
	store32(w+4, 0)
	store32(w+8, 0)
	store32(w+12, 0)
	store32(w+16, 0)
	store32(w+20, 0)
	store32(w+24, 0)
	store32(w+28, 0)
}

// syncHeapTop republishes the marking range after anything moves the coverage
// line -- a new chunk, or a memory.grow that extends an already-covered slice.
// It is called from BOTH, and forgetting the second was a silent corruption:
// markCandidate accepted addresses in the covered heap that heapTop said were
// outside it, so the sweep freed live objects nothing had marked.
func syncHeapTop() { heapTop = heapBase + coveredSpans()<<spanLog }

// coveredSpans is how much of the heap has metadata behind it, and it -- not
// spanCount -- is what every span loop in this package bounds itself by.
//
// The two differ only between a memory.grow and the growCoverage that follows
// it. A span above this line is backed by real linear memory and is simply not
// heap yet: nothing can be allocated in it, the sweep does not walk it, and the
// next allocation that wants room brings it in. Nothing is ever LOST there,
// which is the difference from the cap this replaces -- initialize() used to
// clamp an adopted memory to HeapCap and drop everything above it on the floor.
func coveredSpans() uint32 {
	c := gcm.chunks << (sliceLog - spanLog)
	if c > gcm.spanCount {
		c = gcm.spanCount
	}
	return c
}

// growCoverage brings one more 4 MiB slice of the heap under metadata, and
// reports whether it could.
//
// WHERE THE CHUNK GOES is the only interesting decision, and it is made to keep
// the TOP of the heap contiguous, because that is where a large object goes.
//
//   - First choice: the LOWEST free run of ten spans in the region that is
//     already covered. A chunk placed there is a permanent blocker in a part of
//     the heap that is already fragmented, and it leaves everything above it in
//     one piece.
//   - Fallback: the first ten spans of the slice being covered, which are
//     guaranteed unassigned because they are above the old coverage line and
//     nothing can have been allocated there. This is the case under real
//     pressure, and it is the one that caps a single object at just under 4 MiB.
//
// agents/sharding.md section 9 proposed a FIXED position at the top of each
// slice. That does not work and the reason is worth keeping: a chunk at the top
// of slice k cannot be written until the heap has grown through the whole slice,
// so a guest wanting 64 KiB of heap would have had to take 4 MiB to describe it.
// Placing it at the bottom is what makes coverage incremental, and the
// "so a run can stretch between chunks" argument for the top is not one --
// consecutive chunks leave a 1,014-span gap either way.
func growCoverage() bool {
	k := gcm.chunks
	if k >= maxChunks {
		return false
	}
	// The slice must be backed at least as far as the chunk itself.
	if gcm.spanCount < k<<(sliceLog-spanLog)+metaChunkSpans {
		return false
	}
	t := scanSpanRun(0, coveredSpans(), metaChunkSpans)
	if t == noSpan {
		t = k << (sliceLog - spanLog)
	}
	addr := heapBase + t<<spanLog
	// ALWAYS wiped, not zero()d: zero() is a no-op under -tags fkgcnozero,
	// which is a measurement arm for allocation cost. A chunk arriving with the
	// last tenant's bytes in it is a bitmap full of set bits and a span table
	// full of invented classes, which is not a measurement arm, it is a heap
	// that walks into somebody else's memory.
	wipe(addr, metaChunkBytes)
	gcm.metaDir[k] = addr
	gcm.chunks = k + 1
	// Only reachable now: setSpanClass for a span inside the new slice needs
	// metaDir[k], and in the fallback case t IS inside the new slice.
	for j := uint32(0); j < metaChunkSpans; j++ {
		setSpanClass(t+j, clsMeta)
	}
	gcm.freeBytes -= metaChunkSpans << spanLog
	gcm.metaSpans += metaChunkSpans
	syncHeapTop()
	return true
}

// MetaBytes is how much linear memory the collector's own metadata occupies,
// and since sharding stage C it is a FUNCTION OF THE HEAP rather than a
// constant.
//
//	MetaBytes = MetaFixedBytes + chunks * 40,960
//	chunks    = ceil(covered heap / 4 MiB)
//
// It is .bss plus heap, so it costs nothing in the wasm binary -- but it IS
// linear memory, and agents/guests.md prices linear memory at 0.2 ms of
// Factorio worst tick per MiB whether or not anything is using it. This number
// belongs in a heap budget rather than hiding in one.
func MetaBytes() uint32 { return MetaFixedBytes() + gcm.chunks*metaChunkBytes }

// MetaFixedBytes is the part that does not scale: the class tables, the mark
// stack, the dirty queue, the directory and the counters. This is what a guest
// pays for having a collector at all, before it allocates anything.
func MetaFixedBytes() uint32 { return uint32(unsafe.Sizeof(gcm)) }

// MetaChunks is how many 4 MiB slices of heap currently have metadata behind
// them, and MetaChunkBytes is what one costs. Exported so that the size model
// is TESTED rather than documented -- see TestTheMetadataSizeModelHolds.
func MetaChunks() uint32     { return gcm.chunks }
func MetaChunkBytes() uint32 { return metaChunkBytes }

// MetaSliceBytes is how much heap one chunk describes.
func MetaSliceBytes() uint32 { return 1 << sliceLog }

File: guest/rust/fkgc/src/meta.rs
//! Collector metadata that GROWS WITH THE HEAP, which is what removes the cap.
//!
//! The mirror of `guest/go/fkgc/meta.go`, and structurally identical to it: the
//! metadata splits by GROWTH LAW rather than by what it means.
//!
//! ```text
//! fixed      the class tables, the mark stack, the free-run heads, the dirty
//!            queue, the counters. ~36 KiB, still .bss, still exactly one
//!            struct (see GcMeta for why that is a correctness requirement).
//! scaling    the mark bitmap, the span class and the span aux. One CHUNK per
//!            4 MiB slice of heap, allocated FROM THE HEAP under the class
//!            CLS_META, reached through a directory that is itself a fixed
//!            .bss array.
//! ```
//!
//! `meta_dir` is 1,024 entries of 4 KiB of `.bss` and that covers the ENTIRE
//! wasm32 address space: 1,024 × 4 MiB is 4 GiB. A flat table sized for 4 GiB
//! would have been 35 MiB of `.bss` -- about 7 ms of Factorio worst tick before
//! the guest allocated a byte -- which is why the directory exists at all.
//!
//! WHY THIS IS SAFE is the one hazard in the design with no precedent elsewhere
//! in the repo, and the answer is the Go collector's: `drain_dirty_spans`
//! already drops a sub-heap card with one compare, and a metadata card inside
//! the heap is dropped by the span-class load `rescan_span` ALREADY performs.
//! The compare moves; it does not multiply. Two things follow and both are
//! enforced rather than assumed -- `mark_candidate` REJECTS a `CLS_META` span,
//! and `meta_dir` holds HEAP ADDRESSES and lives inside `[meta_lo, meta_hi)` so
//! that `scan_roots` subtracts it.

use crate::heap::{
    gcm, load32, scan_span_run, store32, wipe, CLS_META, GRANULE, GRANULE_LOG, NO_SPAN, SPAN_BYTES,
    SPAN_LOG,
};

/// How much heap one chunk describes: 4 MiB.
///
/// It is the whole tuning decision here and it trades two things against each
/// other. Bigger slices mean a smaller directory and a larger FLOOR -- the first
/// chunk is paid by any guest that allocates at all. At 4 MiB the floor is
/// 40 KiB and the directory is 4 KiB.
pub(crate) const SLICE_LOG: u32 = 22;
/// 1,024 spans per chunk.
pub(crate) const SLICE_SPANS: u32 = 1 << (SLICE_LOG - SPAN_LOG);
/// 262,144 granules per chunk.
pub(crate) const SLICE_WORDS: u32 = 1 << (SLICE_LOG - GRANULE_LOG);

// The chunk's layout, in bytes from its base. Three arrays of u32, in the order
// they are hottest.
pub(crate) const META_MARK_OFF: u32 = 0;
/// 32,768 B: one bit per granule.
const META_MARK_LEN: u32 = SLICE_WORDS / 32 * 4;
const META_CLASS_OFF: u32 = META_MARK_OFF + META_MARK_LEN;
/// 4,096 B: one word per span.
const META_CLASS_LEN: u32 = SLICE_SPANS * 4;
const META_AUX_OFF: u32 = META_CLASS_OFF + META_CLASS_LEN;
const META_AUX_LEN: u32 = SLICE_SPANS * 4;

/// 40,960 B is exactly ten spans, with nothing left over. That is not luck -- it
/// is what picking `u32` for all three tables buys, and a layout that did not
/// land on a span boundary would waste the remainder of the last one forever.
pub(crate) const META_CHUNK_BYTES: u32 = META_AUX_OFF + META_AUX_LEN;
pub(crate) const META_CHUNK_SPANS: u32 = META_CHUNK_BYTES >> SPAN_LOG;

/// Covers all of wasm32. See `max_spans` for the bound that actually binds,
/// which is smaller and is about `u32` arithmetic rather than about this.
pub(crate) const MAX_CHUNKS: u32 = 1 << (32 - SLICE_LOG);

/// Eight: a span is 4,096 B, a granule is 16 B, so a span is exactly 256
/// granules and exactly eight bitmap words. The bitmap partitions on span
/// boundaries with nothing left over, which is what lets a sweep clear as it
/// goes and what lets every per-span loop compute ONE address and then index
/// within it.
pub(crate) const MARK_WORDS_PER_SPAN: u32 = SPAN_BYTES / GRANULE / 32;

const _: () = assert!(META_CHUNK_BYTES == META_CHUNK_SPANS * SPAN_BYTES);

/// What span `si` holds: 0 unassigned, `1..=NUM_CLASSES` a size class,
/// `CLS_LARGE`/`CLS_LARGE_MID` a large-object run, `CLS_META` a metadata chunk.
///
/// The value may carry `CLS_FRESH` or `CLS_PENDING`; every reader that cares
/// about the CLASS masks them off, and `sweep_span` is the one reader that wants
/// the raw word.
#[inline(always)]
pub(crate) fn span_class_of(si: u32) -> u32 {
    let g = gcm();
    load32(g.meta_dir[(si >> (SLICE_LOG - SPAN_LOG)) as usize] + META_CLASS_OFF + ((si & (SLICE_SPANS - 1)) << 2))
}

#[inline(always)]
pub(crate) fn set_span_class(si: u32, c: u32) {
    let g = gcm();
    store32(
        g.meta_dir[(si >> (SLICE_LOG - SPAN_LOG)) as usize] + META_CLASS_OFF + ((si & (SLICE_SPANS - 1)) << 2),
        c,
    )
}

/// The run length for a `CLS_LARGE` head and the head's span index for a
/// `CLS_LARGE_MID` continuation. Unused otherwise.
#[inline(always)]
pub(crate) fn span_aux_of(si: u32) -> u32 {
    let g = gcm();
    load32(g.meta_dir[(si >> (SLICE_LOG - SPAN_LOG)) as usize] + META_AUX_OFF + ((si & (SLICE_SPANS - 1)) << 2))
}

#[inline(always)]
pub(crate) fn set_span_aux(si: u32, v: u32) {
    let g = gcm();
    store32(
        g.meta_dir[(si >> (SLICE_LOG - SPAN_LOG)) as usize] + META_AUX_OFF + ((si & (SLICE_SPANS - 1)) << 2),
        v,
    )
}

/// The address of the FIRST of span `si`'s eight bitmap words.
///
/// Every per-span loop in the sweep and the re-scan calls this once and then
/// indexes within the span, which is why moving the bitmap into the heap did not
/// make either of them slower.
#[inline(always)]
pub(crate) fn mark_word_base(si: u32) -> u32 {
    let g = gcm();
    g.meta_dir[(si >> (SLICE_LOG - SPAN_LOG)) as usize]
        + META_MARK_OFF
        + (si & (SLICE_SPANS - 1)) * (MARK_WORDS_PER_SPAN * 4)
}

/// Tests the bit for the object at byte offset `off` within the span whose
/// bitmap words start at `mw`. `off` is a span offset, so it is under 4,096 and
/// the granule index is under 256.
#[inline(always)]
pub(crate) fn is_marked_at(mw: u32, off: u32) -> bool {
    let g = off >> GRANULE_LOG;
    load32(mw + ((g >> 5) << 2)) & (1u32 << (g & 31)) != 0
}

/// Wipes span `si`'s eight bitmap words.
///
/// Eight explicit stores rather than a `write_bytes` over 32 bytes: that lowers
/// to `memory.fill`, which the emitter substitutes with the runtime's `mem_fill`
/// -- a call with its own bounds check and page mark, for 32 bytes. The CALL is
/// the cost, not the bytes.
#[inline]
pub(crate) fn clear_span_marks(si: u32) {
    let w = mark_word_base(si);
    store32(w, 0);
    store32(w + 4, 0);
    store32(w + 8, 0);
    store32(w + 12, 0);
    store32(w + 16, 0);
    store32(w + 20, 0);
    store32(w + 24, 0);
    store32(w + 28, 0);
}

/// Republishes the marking range after anything moves the coverage line -- a new
/// chunk, or a `memory.grow` that extends an already-covered slice. It is called
/// from BOTH, and forgetting the second is a silent corruption: `mark_candidate`
/// would accept addresses in the covered heap that `heap_top` said were outside
/// it, so the sweep would free live objects nothing had marked.
#[inline]
pub(crate) fn sync_heap_top() {
    let g = gcm();
    g.heap_top = g.heap_base + (covered_spans() << SPAN_LOG);
}

/// How much of the heap has metadata behind it, and it -- not `span_count` -- is
/// what every span loop in this crate bounds itself by.
///
/// The two differ only between a `memory.grow` and the `grow_coverage` that
/// follows it. A span above this line is backed by real linear memory and is
/// simply not heap yet: nothing can be allocated in it, the sweep does not walk
/// it, and the next allocation that wants room brings it in. Nothing is ever
/// LOST there.
#[inline]
pub(crate) fn covered_spans() -> u32 {
    let g = gcm();
    let c = g.chunks << (SLICE_LOG - SPAN_LOG);
    if c > g.span_count {
        g.span_count
    } else {
        c
    }
}

/// Brings one more 4 MiB slice of the heap under metadata, and reports whether
/// it could.
///
/// WHERE THE CHUNK GOES is the only interesting decision, and it is made to keep
/// the TOP of the heap contiguous, because that is where a large object goes.
///
///   * First choice: the LOWEST free run of ten spans in the region that is
///     already covered. A chunk placed there is a permanent blocker in a part of
///     the heap that is already fragmented, and it leaves everything above it in
///     one piece.
///   * Fallback: the first ten spans of the slice being covered, which are
///     guaranteed unassigned because they are above the old coverage line. This
///     is the case under real pressure, and it is the one that caps a single
///     object at just under 4 MiB.
pub(crate) fn grow_coverage() -> bool {
    let g = gcm();
    let k = g.chunks;
    if k >= MAX_CHUNKS {
        return false;
    }
    // The slice must be backed at least as far as the chunk itself.
    if g.span_count < (k << (SLICE_LOG - SPAN_LOG)) + META_CHUNK_SPANS {
        return false;
    }
    let mut t = scan_span_run(0, covered_spans(), META_CHUNK_SPANS);
    if t == NO_SPAN {
        t = k << (SLICE_LOG - SPAN_LOG);
    }
    let addr = g.heap_base + (t << SPAN_LOG);
    // ALWAYS wiped, never `zero`d: `zero` is a no-op under the `fkgcnozero`
    // measurement arm. A chunk arriving with the last tenant's bytes in it is a
    // bitmap full of set bits and a span table full of invented classes, which is
    // not a measurement arm, it is a heap that walks into somebody else's memory.
    wipe(addr, META_CHUNK_BYTES);
    g.meta_dir[k as usize] = addr;
    g.chunks = k + 1;
    // Only reachable now: `set_span_class` for a span inside the new slice needs
    // meta_dir[k], and in the fallback case t IS inside the new slice.
    let mut j = 0u32;
    while j < META_CHUNK_SPANS {
        set_span_class(t + j, CLS_META);
        j += 1;
    }
    g.free_bytes -= META_CHUNK_SPANS << SPAN_LOG;
    g.meta_spans += META_CHUNK_SPANS;
    sync_heap_top();
    true
}

/// How much linear memory the collector's own metadata occupies, and it is a
/// FUNCTION OF THE HEAP rather than a constant.
///
/// ```text
/// meta_bytes = meta_fixed_bytes + chunks * 40,960
/// chunks     = ceil(covered heap / 4 MiB)
/// ```
///
/// It is `.bss` plus heap, so it costs nothing in the wasm binary -- but it IS
/// linear memory, and `agents/guests.md` prices linear memory at 0.2 ms of
/// Factorio worst tick per MiB whether or not anything is using it.
pub fn meta_bytes() -> u32 {
    meta_fixed_bytes() + gcm().chunks * META_CHUNK_BYTES
}

/// The part that does not scale: the class tables, the mark stack, the dirty
/// queue, the directory and the counters. This is what a guest pays for having a
/// collector at all, before it allocates anything.
pub fn meta_fixed_bytes() -> u32 {
    core::mem::size_of::<crate::heap::GcMeta>() as u32
}

/// How many 4 MiB slices of heap currently have metadata behind them.
pub fn meta_chunks() -> u32 {
    gcm().chunks
}

/// What one chunk costs. Exported so that the size model is TESTED rather than
/// documented.
pub fn meta_chunk_bytes() -> u32 {
    META_CHUNK_BYTES
}

/// How much heap one chunk describes.
pub fn meta_slice_bytes() -> u32 {
    1 << SLICE_LOG
}
